File: typescript/lesson/3-RPG character class hierarchy design/main.ts
/* INTERFACE POUR LES COMPORTEMENS */
interface IDeplacement {
    SeDeplacer() : void;
}

interface IAttaque {
    Attaquer() : void;
}

interface ISoin {
    Soigner() : void;
}

/* IMPLEMENTATION DE TOUT LES COMPORTEMENTS */
class Marcher implements IDeplacement {
    public SeDeplacer(): void {
        console.log("Marcher");
    }
}

class Voler implements IDeplacement {
    public SeDeplacer(): void {
        console.log("Voler");
    }
}

class Courir implements IDeplacement {
    public SeDeplacer(): void {
        console.log("Courir");
    }
}

class PasAttaque implements IAttaque {
    public Attaquer(): void {
        console.log("Pas d'attaque");
    }
}

class Fusil implements IAttaque {
    public Attaquer(): void {
        console.log("Fusil");
    }
}

class AvecSort implements IAttaque {
    public Attaquer(): void {
        console.log("Avec sort");
    }
}

class PasSoin implements ISoin {
    public Soigner(): void {
        console.log("Pas de soin");
    }
}

class SuperSoin implements ISoin {
    public Soigner(): void {
        console.log("Super Soin");
    }
}

class SoinBase implements ISoin {
    public Soigner(): void {
        console.log("Soin de base");
    }
}


/* CLASS PERSONNAGE */
abstract class Personnage {
    protected deplacement: IDeplacement;
    protected attaque: IAttaque;
    protected soin: ISoin;

    constructor(deplacement: IDeplacement, attaque: IAttaque, soin: ISoin){
        this.deplacement = deplacement;
        this.attaque = attaque;
        this.soin = soin;
    }

    public SeDeplacer(): void {
        this.deplacement.SeDeplacer();
    }

    public Attaquer(): void {
        this.attaque.Attaquer();
    }

    public Soigner(): void {
        this.soin.Soigner();
    }
}

class Roi extends Personnage {
    constructor(){
        super(new Marcher(), new PasAttaque(), new PasSoin());
    }
}

class Soldat extends Personnage {
    constructor(){
        super(new Courir(), new Fusil(), new SoinBase());
    }
}

class Sorcier extends Personnage {
    constructor(){
        super(new Voler(), new AvecSort(), new SoinBase());
    }
}

class Pretre extends Personnage {
    constructor(){
        super(new Marcher(), new PasAttaque(), new SuperSoin());
    }
}

function Main() {
    const roi: Personnage = new Roi();
    const pretre: Personnage = new Pretre();
    const soldat: Personnage = new Soldat();
    const sorcier: Personnage = new Sorcier();

    console.log("\n\nRoi: ");
    roi.SeDeplacer();
    roi.Attaquer();
    roi.Soigner();
    
    console.log("\n\nPrêtre: ");
    pretre.SeDeplacer();
    pretre.Attaquer();
    pretre.Soigner();
    
    console.log("\n\nSoldat: ");
    soldat.SeDeplacer();
    soldat.Attaquer();
    soldat.Soigner();
    
    console.log("\n\nSorcier: ");
    sorcier.SeDeplacer();
    sorcier.Attaquer();
    sorcier.Soigner();
}
Main();

File: typescript/lesson/3-RPG character class hierarchy design/main.cs
using System;

public interface IDeplacement {
    void SeDeplacer();
}

public interface IAttaque {
    void Attaquer();
}

public interface ISoin {
    void Soigner();
}

public class Marcher : IDeplacement {
    public void SeDeplacer(){
        Console.WriteLine("Marcher");
    }
}

public class Voler : IDeplacement {
    public void SeDeplacer(){
        Console.WriteLine("Voler");
    }
}

public class Courir : IDeplacement {
    public void SeDeplacer(){
        Console.WriteLine("Courir");
    }
}

public class PasAttaque : IAttaque {
    public void Attaquer() {
        Console.WriteLine("Pas d'attaque");
    }
} 

public class Fusil : IAttaque {
    public void Attaquer() {
        Console.WriteLine("Fusil");
    }
} 

public class AvecSort : IAttaque {
    public void Attaquer() {
        Console.WriteLine("Avec un sort");
    }
} 

public class PasSoin: ISoin {
    public void Soigner(){
        Console.WriteLine("Pas de soin");
    }
}

public class SoinBase : ISoin {
    public void Soigner(){
        Console.WriteLine("Soin de base");
    }
}

public class SuperSoin : ISoin {
    public void Soigner(){
        Console.WriteLine("Super Soin");
    }
}


public abstract class Personnage {
    protected  IDeplacement deplacement;
    protected  IAttaque attaque;
    protected  ISoin soin;

    public Personnage(){}
    public Personnage(IDeplacement deplacement,  IAttaque attaque, ISoin soin){
        this.deplacement = deplacement;
        this.attaque = attaque;
        this.soin = soin;
    }
    
    public void SeDeplacer() {
        this.deplacement.SeDeplacer();
    }

    public void Soigner(){
        this.soin.Soigner();
    }

    public void Attaquer(){
        this.attaque.Attaquer();
    }
}

public class Roi : Personnage {
    public Roi() : base(new Marcher(), new PasAttaque(), new PasSoin()) {}
}

public class Soldat : Personnage {
    public Soldat() : base(
        new Courir(),
        new Fusil(),
        new SoinBase()
    ){}
}

public class Sorcier : Personnage {
    public Sorcier(){
        this.deplacement = new Voler();
        this.attaque = new AvecSort();
        this.soin = new SoinBase();
    }
}

public class Pretre : Personnage {
    public Pretre() : base(
        new Marcher(),
        new PasAttaque(),
        new SuperSoin()
    ){}
}

public static class Program {
    public static void Main() {
        Personnage roi = new Roi();
        Personnage pretre = new Pretre();
        Personnage soldat = new Soldat();
        Personnage sorcier = new Sorcier();

        Console.WriteLine("Roi: ");
        roi.SeDeplacer();
        roi.Attaquer();
        roi.Soigner();
    }
}

